recirq/time_crystals/dtctask.py

```diff
@@ -0,0 +1,254 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cirq
+import recirq
+import datetime
+import itertools
+import numpy as np
+from typing import Sequence, Optional, Dict
+import os
+
+EXPERIMENT_NAME = "time_crystals"
+DEFAULT_BASE_DIR = os.path.expanduser(f'~/cirq_results/{EXPERIMENT_NAME}')
+
+
+@recirq.json_serializable_dataclass(namespace='recirq.readout_scan',
+                                    registry=recirq.Registry,
+                                    frozen=False)
+class CompareDTCTask:
+    """ A task for managing inputs to a comparison Discrete Time Crystal experiment, comparing different options for parameters
+
+    Attributes
+        dataset_id: unique identifier for this dataset
+        qubits: chain of connected qubits available for the circuit
+        cycles: number of DTC cycles to consider for circuits
+        circuit_list: symbolic DTC circuit list
+        disorder_instances: number of disorder instances averaged over
+        options_dict: dict mapping DTCTask attribute names to options for that attribute, to take a product over
+        options_order: sequence of keys in options_dict, defining order of product over options
+    """
+
+    # Task parameters
+    dataset_id: str
+
+    # experiment parameters
+    qubits: Sequence[cirq.Qid]
+    cycles: int
+    disorder_instances: int
+    circuit_list: Sequence[cirq.Circuit]
+
+    # options to take product over
+    options_dict: Dict[str, Sequence[np.ndarray]]
+    options_order: Sequence[str]
+
+    def __init__(
+            self,
+            qubits: Sequence[cirq.Qid],
+            cycles: int,
+            disorder_instances: int,
+            options_dict: Dict[str, Sequence[np.ndarray]],
+            options_order: Optional[Sequence[str]] = None):
+
+        self.dataset_id = datetime.datetime.utcnow()
+
+        self.qubits = qubits
+        self.cycles = cycles
+        self.disorder_instances = disorder_instances
+
+        # create symbolic circuit list from qubits and cycles count
+        self.circuit_list = recirq.time_crystals.symbolic_dtc_circuit_list(qubits, cycles)
+
+        self.options_dict = options_dict
+        self.options_order = list(self.options_dict.keys()) if options_order is None else options_order
+
+        # check that the input parameters are consistent
+        assert set(self.options_order) == set(self.options_dict.keys()), 'options_order and the keys of options_dict are not the same'
+        assert not {'initial_states', 'initial_state'} <= self.options_dict.keys(), 'do not supply both initial_states and initial_state'
+
+
+    @property
+    def fn(self):
+        fn = (f'{self.dataset_id}/'
+                f'{len(self.qubits)}/'
```

Comment:
are you ok with any task with the same number of qubits but different qubit identities mapping to the same thing?